### PR TITLE
Improve the API to access vulnerabilities

### PR DESCRIPTION
**Pull request by oheger-bosch**
AdvisorRecord: Add a getVulnerabilities() function

The function allows obtaining all the vulnerabilities for a specific
package independent on the advisor that produced the result. If the
same vulnerabilities have been retrieved from different advisors (with
potentially different metadata), their references are merged.

This can be useful for rules that operate on advisor results.

### CODE REVIEW

File: model/src/main/kotlin/AdvisorRecord.kt [hunk 1]
```diff
@@ -61,6 +61,32 @@ data class AdvisorRecord(
             results.any { it.summary.issues.isNotEmpty() }
         }
     }
+
+    /**
+     * Return a list with all [Vulnerability] objects that have been found for the given [package][pkgId]. Results
+     * from different advisors are merged if necessary.
+     */
+    fun getVulnerabilities(pkgId: Identifier): List<Vulnerability> {
+        return advisorResults[pkgId]?.flatMap { it.vulnerabilities }?.mergeVulnerabilities() ?: emptyList()
```
Comment (sschuberth) [inline model/src/main/kotlin/AdvisorRecord.kt:70]: Nit: Could be a function expression. Also, does `orEmpty()` work here?

File: model/src/main/kotlin/AdvisorRecord.kt [hunk 2]
```diff
@@ -61,6 +61,32 @@ data class AdvisorRecord(
...
+    }
+}
+
+/**
+ * Merge this list of [Vulnerability] objects by combining vulnerabilities with the same ID and merging their
+ * references.
+ */
+private fun List<Vulnerability>.mergeVulnerabilities(): List<Vulnerability> {
+    val vulnerabilitiesById = groupByTo(sortedMapOf()) { it.id }
+    return vulnerabilitiesById.map { entry -> entry.value.mergeReferences() }
```
Comment (sschuberth) [inline model/src/main/kotlin/AdvisorRecord.kt:80]: Nit: For oneliners we usually stick to `it`.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: model/src/test/kotlin/AdvisorRecordTest.kt [hunk 3]
```diff
@@ -66,11 +74,139 @@ class AdvisorRecordTest : StringSpec({
...
+            VulnerabilityReference(URI("$uriPrefix$id"), scoringSystem, severity)
+        )
+    )
+
+/**
+ * Create an [AdvisorResult] for an advisor with the given [advisorIdx] with the passed in [issues] and
+ * [vulnerabilities].
+ */
+private fun createResult(
+    advisorIdx: Int,
```
Comment (MarcelBochtler) [inline model/src/test/kotlin/AdvisorRecordTest.kt:275]: what is the reason to call this parameter `advisorIdx` instead of `advisorId`?

Comment (oheger-bosch) [reply model/src/test/kotlin/AdvisorRecordTest.kt:275]: It is a numeric index, which is then used to generate a synthetic advisor ID, such as `advisor1`, `advisor2`, etc.

File: model/src/test/kotlin/AdvisorRecordTest.kt [hunk 4]
```diff
@@ -66,11 +74,139 @@ class AdvisorRecordTest : StringSpec({
...
+            record.advisorResults should haveSize(1)
+            record.advisorResults.entries.single().let { (id, results) ->
+                id shouldBe Identifier("type:namespace:name:version")
+                results shouldHaveSize 1
+            }
+        }
+    }
+
+    "hasIssues" should {
+        "detect a result that has no issues" {
```
Comment (fviernau) [inline model/src/test/kotlin/AdvisorRecordTest.kt:87]: consider: "return false given the record has no issues"

File: model/src/test/kotlin/AdvisorRecordTest.kt [hunk 5]
```diff
@@ -66,11 +74,139 @@ class AdvisorRecordTest : StringSpec({
             has_issues: false
         """.trimIndent()
 
-        val record = shouldNotThrowAny { yamlMapper.readValue<AdvisorRecord>(yaml) }
-        record.advisorResults should haveSize(1)
-        record.advisorResults.entries.single().let { (id, results) ->
-            id shouldBe Identifier("type:namespace:name:version")
-            results shouldHaveSize 1
+            val record = shouldNotThrowAny { yamlMapper.readValue<AdvisorRecord>(yaml) }
+            record.advisorResults should haveSize(1)
```
Comment (fviernau) [inline model/src/test/kotlin/AdvisorRecordTest.kt:78]: consider:

```
with (record.advisorResults) {
   should haveSize(1)
   ...
}
```

File: model/src/test/kotlin/AdvisorRecordTest.kt [hunk 6]
```diff
@@ -66,11 +74,139 @@ class AdvisorRecordTest : StringSpec({
...
+ */
+private fun createResult(
+    advisorIdx: Int,
+    issues: List<OrtIssue> = emptyList(),
+    vulnerabilities: List<Vulnerability> = emptyList()
+): AdvisorResult {
+    val details = AdvisorDetails("advisor$advisorIdx")
+    val summary = AdvisorSummary(
+        startTime = Instant.parse("2021-04-06T13:26:05.123Z"),
+        endTime = Instant.parse("2021-04-06T13:26:47.456Z"), issues = issues
```
Comment (fviernau) [inline model/src/test/kotlin/AdvisorRecordTest.kt:282]: PLease move `issues = issues` to a separate line.

File: model/src/test/kotlin/AdvisorRecordTest.kt [hunk 7]
```diff
@@ -22,15 +22,23 @@ package org.ossreviewtoolkit.model
 import com.fasterxml.jackson.module.kotlin.readValue
 
 import io.kotest.assertions.throwables.shouldNotThrowAny
-import io.kotest.core.spec.style.StringSpec
+import io.kotest.core.spec.style.WordSpec
```
Comment (fviernau) [inline model/src/test/kotlin/AdvisorRecordTest.kt:25]: Do you want to amend the Copyright?

File: model/src/main/kotlin/AdvisorRecord.kt (hunk 1 above)
Comment (fviernau) [reply model/src/main/kotlin/AdvisorRecord.kt:70]: starting with `return advisorResults[pkgId].orEmpty().flatMap...` should make it a bit more compact.

File: model/src/main/kotlin/AdvisorRecord.kt [hunk 8]
```diff
@@ -61,6 +61,32 @@ data class AdvisorRecord(
...
+    fun getVulnerabilities(pkgId: Identifier): List<Vulnerability> {
+        return advisorResults[pkgId]?.flatMap { it.vulnerabilities }?.mergeVulnerabilities() ?: emptyList()
+    }
+}
+
+/**
+ * Merge this list of [Vulnerability] objects by combining vulnerabilities with the same ID and merging their
+ * references.
+ */
+private fun List<Vulnerability>.mergeVulnerabilities(): List<Vulnerability> {
```
Comment (fviernau) [inline model/src/main/kotlin/AdvisorRecord.kt:78]: Please make this a `private fun Collection<Vulnerability> ...` .

File: model/src/main/kotlin/AdvisorRecord.kt [hunk 9]
```diff
@@ -61,6 +61,32 @@ data class AdvisorRecord(
...
+    val vulnerabilitiesById = groupByTo(sortedMapOf()) { it.id }
+    return vulnerabilitiesById.map { entry -> entry.value.mergeReferences() }
+}
+
+/**
+ * Merge this (non-empty) list of [Vulnerability] objects (which are expected to have the same ID) by to a single
+ * [Vulnerability] that contains all the references from the source vulnerabilities (with duplicates removed).
+ */
+private fun List<Vulnerability>.mergeReferences(): Vulnerability {
+    val references = flatMapTo(mutableSetOf()) { it.references }
```
Comment (fviernau) [inline model/src/main/kotlin/AdvisorRecord.kt:88]: Please make this a `private fun Collection<Vulnerability> ...` .

**Review by fviernau — CHANGES_REQUESTED** [review]

File: model/src/test/kotlin/AdvisorRecordTest.kt [hunk 10]
```diff
@@ -66,11 +74,142 @@ class AdvisorRecordTest : StringSpec({
             has_issues: false
         """.trimIndent()
 
-        val record = shouldNotThrowAny { yamlMapper.readValue<AdvisorRecord>(yaml) }
-        record.advisorResults should haveSize(1)
-        record.advisorResults.entries.single().let { (id, results) ->
-            id shouldBe Identifier("type:namespace:name:version")
-            results shouldHaveSize 1
+            val record = shouldNotThrowAny { yamlMapper.readValue<AdvisorRecord>(yaml) }
+            with(record.advisorResults) {
```
Comment (sschuberth) [inline model/src/test/kotlin/AdvisorRecordTest.kt:79]: Nit: It seems a bit odd to use `with` here but not above. But actually, IMO using `with` does not pay off here really, and something like `this should haveSize(1)` reads a bit off to me.

File: model/src/test/kotlin/AdvisorRecordTest.kt (hunk 3 above)
Comment (sschuberth) [reply model/src/test/kotlin/AdvisorRecordTest.kt:278]: As a side note, I'm not a big fan of the "Idx" abbreviation. It's only 2 chars shorter than "Index" and, as we saw where, it can be misread as "id-x".

File: model/src/main/kotlin/AdvisorRecord.kt [hunk 11]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2020 Bosch.IO GmbH
+ * Copyright (C) 2020-2021 Bosch.IO GmbH
```
Comment (sschuberth) [inline model/src/main/kotlin/AdvisorRecord.kt:2]: Commit message: s/independent on the advisor/independent of the advisor/

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: model/src/test/kotlin/AdvisorRecordTest.kt (hunk 10 above)
Comment (oheger-bosch) [reply model/src/test/kotlin/AdvisorRecordTest.kt:79]: There seems to be no agreement between the reviewers here. I just want to note that this piece of code has not actually been changed by this PR. It appears in the diff because I have changed the test class from extending `StringSpec` to `WordSpec`, and therefore indentation changed. So should I revert to the original code?

File: model/src/main/kotlin/AdvisorRecord.kt (hunk 11 above)
Comment (oheger-bosch) [reply model/src/main/kotlin/AdvisorRecord.kt:2]: The parameter is now named `advisorIndex`. Fixed the commit message.

File: model/src/test/kotlin/AdvisorRecordTest.kt [hunk 12]
```diff
@@ -84,7 +85,7 @@ class AdvisorRecordTest : WordSpec({
     }
 
     "hasIssues" should {
-        "detect a result that has no issues" {
+        "return false given the record has no issues" {
```
Comment (sschuberth) [inline model/src/test/kotlin/AdvisorRecordTest.kt:88]: Nit: Strictly speaking that's an unrelated change that should go to a separate commit. But let's turn a blind eye on that for now.

**Review by sschuberth — APPROVED** [review]

File: model/src/test/kotlin/AdvisorRecordTest.kt [hunk 13]
```diff
@@ -97,7 +98,7 @@ class AdvisorRecordTest : WordSpec({
             record.hasIssues shouldBe false
         }
 
-        "detect a result that has issues" {
+        "return true given the record has issues" {
```
Comment (sschuberth) [inline model/src/test/kotlin/AdvisorRecordTest.kt:101]: Please move this fixup to the first commit in this PR, which introduced this line.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by sschuberth — APPROVED** [review]
